text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+
+```js
+const Parent_method_target = Parent.prototype.method;
+
+export function __wasm_bindgen_Parent_method(idx) {
+    Parent_method_target.call(getObject(idx));
+}
+```
+
+If we assume for a moment that [`anyref` is implemented][reference-types] we
+could instead change this to:
+
+```js
+const Parent_method_target = Parent.prototype.method;
+
+export function __wasm_bindgen_Parent_method(obj) {
+    Parent_method_target.call(obj);
+}
+```
+
+(note the lack of need for `getObject`). And finally, with [host bindings] we
+can say that the wasm module's import of `__wasm_bindgen_Parent_method` uses the
+first parameter as `this`, meaning we can transform this to:
+
+```js
+export const __wasm_bindgen_Parent_method = Parent.prototype.method;
+```
+
+and *voila*, no JS shims necessary! With `structural` we'll still need a shim in
+this future world:
+
+```js
+export const __wasm_bindgen_Parent_method = function() { this.method(); };
+```
+
+Alright, with some of those basics out of the way, let's get back to
+why-`final`-by-default. The promise of [host bindings] is that by eliminating
+all these JS shims necessary we can be faster than we would otherwise be,
+providing a feeling that `final` is faster than `structural`. This future,
+however, relies on a number of unimplemented features in wasm engines today.
+Let's consequently get an idea of what the performance looks like today!
+
+I've been slowly over time preparing a [microbenchmark suite][bm] for measuring
+JS/wasm/wasm-bindgen performance. The interesting one here is the benchmark
+"`structural` vs not". If you click "Run test" in a browser after awhile you'll
+see two bars show up. The left-hand one is a method call with `final` and the
+right-hand one is a method call with `structural`. The results I see on my
+computer are:
+
+* Firefox 62, `structural` is 3% faster
+* Firefox 64, `structural` is 3% slower
+* Chrome 69, `structural` is 5% slower
+* Edge 42, `structural` is 22% slower
+* Safari 12, `strutural` is 17% slower
+
+So it looks like for Firefox/Chrome it's not really making much of a difference
+but in Edge/Safari it's much faster to use `final`! It turns out, however, that
+we're not optimizing `structural` as much as we can. Let's change our generated
+code from:
+
+```js
+const Parent_method_target = function() { this.method(); };
+
+export function __wasm_bindgen_Parent_method(obj) {
+    Parent_method_target.call(getObject(obj));
+}
+```
+
+to...
+
+```js
+export function __wasm_bindgen_Parent_method(obj) {
+    getObject(obj).method();
+}
+```
+
+(manually editing the JS today)
+
+and if we rerun the benchmarks (sorry no online demo) we get:
+
+* Firefox 62, `structural` is 22% faster
+* Firefox 64, `structural` is 10% faster
+* Chrome 69, `structural` is 0.3% slower
+* Edge 42, `structural` is 15% faster
+* Safai 12, `structural` is 8% slower
+
+and these numbers look quite different! There's some strong data here showing
```

Comment:
Thanks, this is good data. It also aligns with my personal beliefs about JS engines: because `structural` is ***much*** more common in JS than `final`, it makes sense that the JS engines heavily optimize for `structural`.